forch/radius_query.py

```diff
@@ -27,6 +29,14 @@ def __init__(self, socket_info, radius_secret):
         self.radius_socket = RadiusSocket(socket_info.listen_ip, socket_info.listen_port,
                                           socket_info.server_ip, socket_info.server_port)
         self.radius_socket.setup()
+        self.radius_socket.lock = RLock()
```

Comment:
what is this for?  a) using .lock is horrific because you're breaking encapsulation, and b) I really don't think it's necessary. How is sending a UDP packet require a lock?

Reply:
Python socket isn't thread safe. The alternative would be having a thread safe queue into which state machines dump their requests into and a worker thread keeps emptying it.

Reply:
MAde changes to isolate lock in socket. Is it ok?

Reply:
yes, looks about right!